pkg/kubectl/cmd/create_service.go

```diff
@@ -0,0 +1,214 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/spf13/cobra"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/kubectl"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+)
+
+// NewCmdCreateService is a macro command to create a new namespace
+func NewCmdCreateService(f *cmdutil.Factory, cmdOut io.Writer) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "service",
+		Short: "Create a service using specified subcommand.",
+		Long:  "Create a service using specified subcommand.",
+		Run: func(cmd *cobra.Command, args []string) {
+			cmd.Help()
+		},
+	}
+	cmd.AddCommand(NewCmdCreateServiceClusterIP(f, cmdOut))
+	cmd.AddCommand(NewCmdCreateServiceNodePort(f, cmdOut))
+	cmd.AddCommand(NewCmdCreateServiceLoadBalancer(f, cmdOut))
+
+	return cmd
+}
+
+// ---------------
+
+const (
+	serviceClusterIPLong = `
+Create a clusterIP service with the specified name.`
+
+	serviceClusterIPExample = `  # Create a new clusterIP service named my-cs
+  kubectl create service clusterip my-cs --tcp=5678:8080`
+)
+
+func addPortFlags(cmd *cobra.Command) {
+	cmd.Flags().StringSlice("tcp", []string{}, "Port pairs can be specified as '<port>:<targetPort>'.")
+}
+
+// NewCmdCreateServiceClusterIP is a command to create generic secrets from files, directories, or literal values
+func NewCmdCreateServiceClusterIP(f *cmdutil.Factory, cmdOut io.Writer) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "clusterip NAME [--tcp=<port>:<targetPort>] [--dry-run]",
+		Short:   "Create a clusterIP service.",
+		Long:    serviceClusterIPLong,
+		Example: serviceClusterIPExample,
+		Run: func(cmd *cobra.Command, args []string) {
+			err := CreateServiceClusterIP(f, cmdOut, cmd, args)
+			cmdutil.CheckErr(err)
+		},
+	}
+	cmdutil.AddApplyAnnotationFlags(cmd)
+	cmdutil.AddValidateFlags(cmd)
+	cmdutil.AddPrinterFlags(cmd)
+	cmdutil.AddGeneratorFlags(cmd, cmdutil.ServiceClusterIPGeneratorV1Name)
+	addPortFlags(cmd)
+	cmd.Flags().Bool("headless", false, "Make this service 'headless' (do not do loadbalancing).")
```

Comment:
I'm really hesitant to enshrine "headless" in an argument - our use up to
this point has entirely been colloquial and not connected to the API, and
no other use refers to it as that.  Since you can only set cluster IP on
creation, a string flag "--cluster-ip=[<IP>|None]" with a good description
would be better than inventing a new term.  The description of the command
can describe what --cluster-ip=None does.

On Sun, Jun 26, 2016 at 4:36 AM, Angus Salkeld <notifications@github.com>
wrote:

> In pkg/kubectl/cmd/create_service.go
> <https://github.com/kubernetes/kubernetes/pull/27109#discussion_r68499495>
> :
>
> > +	cmd := &cobra.Command{
> > +		Use:     "clusterip NAME [--tcp=<port>:<targetPort>] [--dry-run]",
> > +		Short:   "Create a clusterIP service.",
> > +		Long:    serviceClusterIPLong,
> > +		Example: serviceClusterIPExample,
> > +		Run: func(cmd *cobra.Command, args []string) {
> > +			err := CreateServiceClusterIP(f, cmdOut, cmd, args)
> > +			cmdutil.CheckErr(err)
> > +		},
> > +	}
> > +	cmdutil.AddApplyAnnotationFlags(cmd)
> > +	cmdutil.AddValidateFlags(cmd)
> > +	cmdutil.AddPrinterFlags(cmd)
> > +	cmdutil.AddGeneratorFlags(cmd, cmdutil.ServiceClusterIPGeneratorV1Name)
> > +	addPortFlags(cmd)
> > +	cmd.Flags().Bool("headless", false, "Make this service 'headless' (do not do loadbalancing).")
>
> @deads2k <https://github.com/deads2k> here is the headless flag, I'll add
> it to the example text to make it more obvious.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/27109/files/a3d7f495b814f2dd5a595ea1a8f4879ebfd9828a#r68499495>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_p5S87x19y1PpV2IXty8zV1pql4zfks5qPjoVgaJpZM4Ixl7H>
> .
>